File: Java/li2/plp/imperative2/Exemplo2.java
package li2.plp.imperative2;

import li2.plp.expressions1.util.TipoPrimitivo;
import li2.plp.expressions2.expression.ExpSoma;
import li2.plp.expressions2.expression.Id;
import li2.plp.expressions2.expression.ValorInteiro;
import li2.plp.imperative1.command.Atribuicao;
import li2.plp.imperative1.command.ComandoDeclaracao;
import li2.plp.imperative1.command.SequenciaComando;
import li2.plp.imperative1.command.Write;
import li2.plp.imperative1.declaration.DeclaracaoComposta;
import li2.plp.imperative1.declaration.DeclaracaoVariavel;
import li2.plp.imperative1.memory.ContextoCompilacaoImperativa;
import li2.plp.imperative1.memory.ListaValor;
import li2.plp.imperative2.command.ChamadaProcedimento;
import li2.plp.imperative2.command.ListaExpressao;
import li2.plp.imperative2.declaration.DeclaracaoParametro;
import li2.plp.imperative2.declaration.DeclaracaoProcedimento;
import li2.plp.imperative2.declaration.DefProcedimento;
import li2.plp.imperative2.declaration.ListaDeclaracaoParametro;
import li2.plp.imperative2.memory.ContextoExecucaoImperativa2;

/**
 * Exemplo2 - Executa o seguinte programa na linguagem li2:
 *
 * { var x = 0, proc p (int y) { x := x + y };
 *   { var x = 1; call p(3); write(x) };
 *   call p(4); write(x)
 * }
 */
public class Exemplo2 {

    /**
     * Executa um programa li2, verificando tipos antes de rodar.
     * Imprime a saida produzida pelo write ou a excecao capturada.
     *
     * @param descricao texto descritivo do programa
     * @param prog      o programa a executar
     * @param ambiente  o ambiente de execucao (contexto imperativo2)
     */
    static void executar(String descricao, Programa prog,
                         ContextoExecucaoImperativa2 ambiente) {
        System.out.println("--------------------------------------------------");
        System.out.println("Programa: \n" + descricao + "\n");
        try {
            boolean bemTipado = prog.checaTipo(new ContextoCompilacaoImperativa(new ListaValor()));
            System.out.println("Bem tipado: " + bemTipado);
            if (bemTipado) {
                ListaValor saida = prog.executar(ambiente);
                System.out.print("Saida: ");
                while (saida.length() > 0) {
                    System.out.print(saida.getHead() + " ");
                    saida = (ListaValor) saida.getTail();
                }
                System.out.println();
            } else {
                System.out.println("Erro de tipo: execucao abortada.");
            }
        } catch (Exception e) {
            System.out.println("Excecao capturada: " + e.getMessage());
            e.printStackTrace();
        }
        System.out.println("--------------------------------------------------");
    }

    public static void main(String[] args) {

        // Identificadores
        Id idX = new Id("x");
        Id idY = new Id("y");
        Id idP = new Id("p");

        // Declaracao de variavel do bloco externo: var x = 0
        DeclaracaoVariavel declX = new DeclaracaoVariavel(idX, new ValorInteiro(0));

        // Parametro formal do procedimento p: int y
        DeclaracaoParametro paramY = new DeclaracaoParametro(idY, TipoPrimitivo.INTEIRO);
        ListaDeclaracaoParametro listaParamY = new ListaDeclaracaoParametro(paramY);

        // Corpo do procedimento p: x := x + y
        Atribuicao corpoP = new Atribuicao(idX, new ExpSoma(idX, idY));

        // Definicao do procedimento p (com parametro formal int y)
        DefProcedimento defP = new DefProcedimento(listaParamY, corpoP);

        // Declaracao de procedimento: proc p (int y) { x := x + y }
        DeclaracaoProcedimento declP = new DeclaracaoProcedimento(idP, defP);

        // Declaracao composta do bloco externo: var x = 0, proc p (int y) { ... }
        DeclaracaoComposta declExterno = new DeclaracaoComposta(declX, declP);

        // Bloco interno: { var x = 1; call p(3); write(x) }
        DeclaracaoVariavel declXInterno = new DeclaracaoVariavel(idX, new ValorInteiro(1));

        // call p(3) — parametro real: 3
        ChamadaProcedimento chamadaP3 = new ChamadaProcedimento(
                idP, new ListaExpressao(new ValorInteiro(3)));

        // write(x) dentro do bloco interno
        Write writeXInterno = new Write(idX);

        // Sequencia interna: call p(3); write(x)
        SequenciaComando seqInterna = new SequenciaComando(chamadaP3, writeXInterno);

        // Bloco interno completo: { var x = 1; call p(3); write(x) }
        ComandoDeclaracao blocoInterno = new ComandoDeclaracao(declXInterno, seqInterna);

        // Comandos do bloco externo apos as declaracoes:
        //   { var x = 1; call p(3); write(x) }; call p(4); write(x)

        // call p(4) — parametro real: 4
        ChamadaProcedimento chamadaP4 = new ChamadaProcedimento(
                idP, new ListaExpressao(new ValorInteiro(4)));

        // write(x) no bloco externo
        Write writeXExterno = new Write(idX);

        // Sequencia externa: blocoInterno; call p(4); write(x)
        SequenciaComando seqExterna = new SequenciaComando(
                blocoInterno,
                new SequenciaComando(chamadaP4, writeXExterno)
        );

        // Bloco completo:
        // { var x = 0, proc p (int y) { x := x + y };
        //   { var x = 1; call p(3); write(x) };
        //   call p(4); write(x)
        // }
        ComandoDeclaracao blocoCompleto = new ComandoDeclaracao(declExterno, seqExterna);

        // Ambiente de execucao
        ContextoExecucaoImperativa2 ambiente =
                new ContextoExecucaoImperativa2(new ListaValor());

        // Programa
        Programa prog = new Programa(blocoCompleto);

        // Executa
        executar(
            "{ var x = 0, proc p (int y) { x := x + y };\n" +
            "    { var x = 1; call p(3); write(x) };\n" +
            "call p(4); write(x) }",
            prog,
            ambiente
        );
    }
}

File: Python/li2-python/Exemplo2.py
import traceback
from util.tipo_primitivo import TipoPrimitivo
from expression.exp_soma import ExpSoma
from expression.id import Id
from expression.valor_inteiro import ValorInteiro
from command.atribuicao import Atribuicao
from command.comando_declaracao import ComandoDeclaracao
from command.sequencia_comando import SequenciaComando
from command.write import Write
from declaration.declaracao_composta import DeclaracaoComposta
from declaration.declaracao_variavel import DeclaracaoVariavel
from memory.contexto_compilacao_imperativa import ContextoCompilacaoImperativa
from memory.lista_valor import ListaValor
from command.chamada_procedimento import ChamadaProcedimento
from command.lista_expressao import ListaExpressao
from declaration.declaracao_parametro import DeclaracaoParametro
from declaration.declaracao_procedimento import DeclaracaoProcedimento
from declaration.def_procedimento import DefProcedimento
from declaration.lista_declaracao_parametro import ListaDeclaracaoParametro
from memory.contexto_execucao_imperativa2 import ContextoExecucaoImperativa2
from Programa import Programa

def executar(descricao, prog, ambiente):
    print("-" * 50)
    print("Programa: \n" + descricao + "\n")
    try:
        bem_tipado = prog.checaTipo(ContextoCompilacaoImperativa(ListaValor()))
        print("Bem tipado:", bem_tipado)
        if bem_tipado:
            saida = prog.executar(ambiente)
            print("Saida: ", end="")
            while saida is not None and saida.length() > 0:
                print(saida.getHead(), end=" ")
                saida = saida.getTail()
            print()
        else:
            print("Erro de tipo: execucao abortada.")
    except Exception as e:
        print("Excecao capturada:", e)
        traceback.print_exc()
    print("-" * 50)

if __name__ == "__main__":
    id_x = Id("x")
    id_y = Id("y")
    id_p = Id("p")

    decl_x = DeclaracaoVariavel(id_x, ValorInteiro(0))
    param_y = DeclaracaoParametro(id_y, TipoPrimitivo.inteiro())
    lista_param_y = ListaDeclaracaoParametro(param_y)

    corpo_p = Atribuicao(id_x, ExpSoma(id_x, id_y))
    def_p = DefProcedimento(lista_param_y, corpo_p)
    decl_p = DeclaracaoProcedimento(id_p, def_p)
    decl_externo = DeclaracaoComposta(decl_x, decl_p)

    decl_x_interno = DeclaracaoVariavel(id_x, ValorInteiro(1))
    chamada_p3 = ChamadaProcedimento(id_p, ListaExpressao(ValorInteiro(3)))
    write_x_interno = Write(id_x)
    seq_interna = SequenciaComando(chamada_p3, write_x_interno)
    bloco_interno = ComandoDeclaracao(decl_x_interno, seq_interna)

    chamada_p4 = ChamadaProcedimento(id_p, ListaExpressao(ValorInteiro(4)))
    write_x_externo = Write(id_x)
    seq_externa = SequenciaComando(bloco_interno, SequenciaComando(chamada_p4, write_x_externo))

    bloco_completo = ComandoDeclaracao(decl_externo, seq_externa)
    ambiente = ContextoExecucaoImperativa2(ListaValor())
    prog = Programa(bloco_completo)

    executar(
        "{ var x = 0, proc p (int y) { x := x + y };\n" +
        "    { var x = 1; call p(3); write(x) };\n" +
        "call p(4); write(x) }",
        prog,
        ambiente
    )
